cmake/modules/extensions.cmake

```diff
@@ -1281,6 +1281,20 @@ function(zephyr_linker_sources location)
   endforeach()
 endfunction(zephyr_linker_sources)
 
+function(zephyr_library_relocate location)
```

Comment:
instead of creating a new function for this, which needs additional processing in the toplevel CMakeLists.txt file, I would much prefer extending the existing `zephyr_code_relocate()` function
https://github.com/zephyrproject-rtos/zephyr/blob/04684c2837dd8af952aff85885c3be7131c9ae37/cmake/modules/extensions.cmake#L1302

to accept generator expressions.

That avoids having one more function to maintain, and keeps other code clean.

Users can then do something like:
```
zephyr_code_relocate($<TARGET_PROPERTY:library,SOURCES> location)
```

A genex is not configure time processing dependent, making the code itself more robust (generally speaking).

As a small benefit, it will even allow downstream users to relocate files in a Zephyr library without patching Zephyr itself.

Reply:
I've been looking into this, and I'm not sure how we can generically support CMake generator expressions in this function. If there's a CMake feature I'm unaware of you think might enable this, please let me know.

The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly, since this is what the `gen_relocate_app.py` script will do internally. This would require API changes for the `zephyr_code_relocate` macro though, which I'd prefer to avoid.

Reply:
> The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Easy.
Add this code to any sample CMakeLists.txt and try to run `ninja kernel_sources`:
```
add_custom_target(kernel_sources COMMAND ${CMAKE_COMMAND} -E echo $<TARGET_PROPERTY:kernel,SOURCE_DIR>/$<JOIN:$<TARGET_PROPERTY:kernel,SOURCES>,$<COMMA>$<TARGET_PROPERTY:kernel,SOURCE_DIR>/> )
```

That will print:
```
/projects/github/ncs/zephyr/kernel/main_weak.c,/projects/github/ncs/zephyr/kernel/banner.c,/projects/github/ncs/zephyr/kernel/device.c,/projects/github/ncs/zephyr/kernel/errno.c,/projects/github/ncs/zephyr/kernel/fatal.c,/projects/github/ncs/zephyr/kernel/init.c,/projects/github/ncs/zephyr/kernel/kheap.c,/projects/github/ncs/zephyr/kernel/mem_slab.c,/projects/github/ncs/zephyr/kernel/thread.c,/projects/github/ncs/zephyr/kernel/version.c,/projects/github/ncs/zephyr/kernel/idle.c,/projects/github/ncs/zephyr/kernel/mailbox.c,/projects/github/ncs/zephyr/kernel/msg_q.c,/projects/github/ncs/zephyr/kernel/mutex.c,/projects/github/ncs/zephyr/kernel/queue.c,/projects/github/ncs/zephyr/kernel/sem.c,/projects/github/ncs/zephyr/kernel/stack.c,/projects/github/ncs/zephyr/kernel/system_work_q.c,/projects/github/ncs/zephyr/kernel/work.c,/projects/github/ncs/zephyr/kernel/sched.c,/projects/github/ncs/zephyr/kernel/condvar.c,/projects/github/ncs/zephyr/kernel/xip.c,/projects/github/ncs/zephyr/kernel/timeout.c,/projects/github/ncs/zephyr/kernel/timer.c,/projects/github/ncs/zephyr/kernel/poll.c,/projects/github/ncs/zephyr/kernel/mempool.c
```

Absolute path to all C files included in the kernel library, separated by a comma.

We probably want to provide an easy to use API for the users for the repeated pattern of including a library.
But if we support any generator expression, then we can easily create an easy to use flag.

For example:
```
zephyr_code_relocate(FILES <list of files or generator expression> LOCATION <location>)
```
but even allow a easy to use argument to the specific library case, which then will do a complex generator expression behind the scenes:
```
zephyr_code_relocate(LIBRARY <lib> LOCATION <location>)
```

> Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly

With generic genex support, anything is possible.
The user of the function decide which expressions to pass to the function.

> This would require API changes for the zephyr_code_relocate macro though, which I'd prefer to avoid.

Fair point, but considering this PR and #49454, then I believe it's time to cleanup the API for this function and ensure it can handle more use-cases.

Reply:
@danieldegrasse if you're not in a hurry for this, i will be happy to take a look at improving the function myself when having the time, but you are of course more than welcome to play around with this yourself.

Reply:
I've updated the `zephyr_code_relocate()` macro based on your suggestions above. Thanks for the pointers- my knowledge on CMake is a bit lacking (especially regarding generator expressions).